File: typescript/tests/healthcare_board.test.ts
// healthcare_board.test.ts
// Verifies the CircleAI.Healthcare port: patient registry, appointment
// scheduling + status update, prescriptions, and the ordering guarantees.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryHealthcareBoard,
  HealthcareDomainContext,
  patient,
  healthAppointment,
  prescription,
} from "../src/healthcare/index";

describe("InMemoryHealthcareBoard", () => {
  it("registers and retrieves patients; unknown id is undefined", () => {
    const b = new InMemoryHealthcareBoard();
    b.register(patient("p1", "Ada", new Date("1990-01-01T00:00:00Z")));
    assert.equal(b.getPatient("p1")?.name, "Ada");
    assert.equal(b.getPatient("nope"), undefined);
  });

  it("schedules appointments and lists them ascending by AtUtc", () => {
    const b = new InMemoryHealthcareBoard();
    b.schedule(healthAppointment("a2", "p1", "Dr B", new Date("2026-03-02T09:00:00Z"), "Booked"));
    b.schedule(healthAppointment("a1", "p1", "Dr A", new Date("2026-03-01T09:00:00Z"), "Booked"));
    b.schedule(healthAppointment("a3", "p2", "Dr C", new Date("2026-01-01T09:00:00Z"), "Booked"));
    const forP1 = b.appointmentsFor("p1");
    assert.deepEqual(
      forP1.map((a) => a.apptId),
      ["a1", "a2"],
    );
  });

  it("updateStatus mutates the stored appointment; unknown throws", () => {
    const b = new InMemoryHealthcareBoard();
    b.schedule(healthAppointment("a1", "p1", "Dr A", new Date("2026-03-01T09:00:00Z"), "Booked"));
    b.updateStatus("a1", "Completed");
    assert.equal(b.appointmentsFor("p1")[0].status, "Completed");
    assert.throws(() => b.updateStatus("ghost", "X"), /Unknown appointment ghost/);
  });

  it("prescriptions list newest-first (PrescribedUtc descending)", () => {
    const b = new InMemoryHealthcareBoard();
    b.prescribe(prescription("r1", "p1", "Med", "1", "od", new Date("2026-01-01T00:00:00Z")));
    b.prescribe(prescription("r2", "p1", "Med2", "2", "bd", new Date("2026-06-01T00:00:00Z")));
    b.prescribe(prescription("r3", "p2", "Med3", "3", "tds", new Date("2026-05-01T00:00:00Z")));
    assert.deepEqual(
      b.prescriptionsFor("p1").map((r) => r.rxId),
      ["r2", "r1"],
    );
  });

  it("null arguments are rejected", () => {
    const b = new InMemoryHealthcareBoard();
    assert.throws(() => b.register(null as never));
    assert.throws(() => b.schedule(null as never));
    assert.throws(() => b.prescribe(null as never));
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(HealthcareDomainContext.systemPromptSnippet.includes("[DOMAIN: Healthcare]"));
    assert.deepEqual(HealthcareDomainContext.complianceFlags, [
      "HIPAA",
      "POPIA",
      "Health_Professions_Act_56_1974",
      "NHA_61_2003",
      "ICD10",
    ]);
    assert.deepEqual(HealthcareDomainContext.suggestedTools, [
      "ehr_system",
      "appointment_scheduler",
      "document_editor",
      "icd10_lookup",
    ]);
  });
});

File: go/tests/healthcare_board_test.go
// healthcare_board_test.go
//
// Verifies the CircleAI.Healthcare port (healthcare_board.go): patient register/
// get, appointment schedule/status-update/ordering (asc), and prescription
// ordering (desc), plus unknown-id errors and deterministic timestamp ties.

package circleai_test

import (
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestHealthcare_PatientRegisterGet(t *testing.T) {
	b := circleai.NewInMemoryHealthcareBoard()
	b.Register(circleai.Patient{PatientId: "p1", Name: "Ada", DateOfBirth: time.Date(1990, 1, 1, 0, 0, 0, 0, time.UTC)})
	if got, ok := b.GetPatient("p1"); !ok || got.Name != "Ada" {
		t.Fatalf("get p1 = %+v ok=%v", got, ok)
	}
	if _, ok := b.GetPatient("none"); ok {
		t.Fatalf("missing patient found")
	}
}

func TestHealthcare_AppointmentsOrderedAndStatus(t *testing.T) {
	b := circleai.NewInMemoryHealthcareBoard()
	base := time.Date(2026, 7, 1, 9, 0, 0, 0, time.UTC)
	b.Schedule(circleai.HealthAppointment{ApptId: "a2", PatientId: "p1", Provider: "Dr X", AtUtc: base.Add(2 * time.Hour), Status: "Booked"})
	b.Schedule(circleai.HealthAppointment{ApptId: "a1", PatientId: "p1", Provider: "Dr X", AtUtc: base, Status: "Booked"})
	b.Schedule(circleai.HealthAppointment{ApptId: "a3", PatientId: "p2", Provider: "Dr Y", AtUtc: base.Add(time.Hour), Status: "Booked"})

	appts := b.AppointmentsFor("p1")
	if len(appts) != 2 || appts[0].ApptId != "a1" || appts[1].ApptId != "a2" {
		t.Fatalf("appointments asc failed: %+v", appts)
	}
	if err := b.UpdateStatus("a1", "Completed"); err != nil {
		t.Fatalf("update status: %v", err)
	}
	if appts = b.AppointmentsFor("p1"); appts[0].Status != "Completed" {
		t.Fatalf("status not updated: %+v", appts[0])
	}
	if err := b.UpdateStatus("ghost", "X"); err == nil {
		t.Fatalf("unknown appointment status update must error")
	}
}

func TestHealthcare_PrescriptionsNewestFirst(t *testing.T) {
	b := circleai.NewInMemoryHealthcareBoard()
	base := time.Date(2026, 6, 1, 0, 0, 0, 0, time.UTC)
	b.Prescribe(circleai.Prescription{RxId: "r1", PatientId: "p1", MedicationName: "A", PrescribedUtc: base})
	b.Prescribe(circleai.Prescription{RxId: "r2", PatientId: "p1", MedicationName: "B", PrescribedUtc: base.Add(48 * time.Hour)})
	b.Prescribe(circleai.Prescription{RxId: "r3", PatientId: "p1", MedicationName: "C", PrescribedUtc: base.Add(24 * time.Hour)})
	rx := b.PrescriptionsFor("p1")
	if len(rx) != 3 || rx[0].RxId != "r2" || rx[1].RxId != "r3" || rx[2].RxId != "r1" {
		t.Fatalf("prescriptions desc failed: %+v", rx)
	}
	if len(b.PrescriptionsFor("other")) != 0 {
		t.Fatalf("other patient should have no prescriptions")
	}
}

func TestHealthcare_DeterministicTies(t *testing.T) {
	same := time.Date(2026, 5, 5, 5, 0, 0, 0, time.UTC)
	for i := 0; i < 5; i++ {
		b := circleai.NewInMemoryHealthcareBoard()
		b.Schedule(circleai.HealthAppointment{ApptId: "c", PatientId: "p", AtUtc: same})
		b.Schedule(circleai.HealthAppointment{ApptId: "a", PatientId: "p", AtUtc: same})
		b.Schedule(circleai.HealthAppointment{ApptId: "b", PatientId: "p", AtUtc: same})
		got := b.AppointmentsFor("p")
		if got[0].ApptId != "a" || got[1].ApptId != "b" || got[2].ApptId != "c" {
			t.Fatalf("iter %d tie order = %v", i, []string{got[0].ApptId, got[1].ApptId, got[2].ApptId})
		}
	}
}
